thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -168,15 +168,65 @@ class CIndexReadSlaveBase : public CSlaveActivity
                 IOutputMetaData *actualFormat = translator ? &translator->queryActualFormat() : expectedFormat;
                 bool tryRemoteStream = actualFormat->queryTypeInfo()->canInterpret() && actualFormat->queryTypeInfo()->canSerialize() &&
                                        projectedFormat->queryTypeInfo()->canInterpret() && projectedFormat->queryTypeInfo()->canSerialize();
-                bool usesBlobs = 0 != (helper->getFlags() & TIRusesblob);
+
+                offset_t avgPartSize = 0;
+                offset_t keyCacheSize = queryJob().getKeyNodeCacheSize() + queryJob().getKeyLeafCacheSize();
+                auto keyCacheRatioHeuristic = [keyCacheSize](offset_t size, unsigned numParts, unsigned compressionMultiple, unsigned cacheSizeMultiple)
+                {
+                    /*
+                     * Rough heuristic.
+                     *
+                     * If ([average compressed part size] * [# parts handling] * [compressionMultiple] < [cacheSizeMultiple]
+                     *
+                     * If the total size of compressed index data * [compressionMultiple - for rough estimate of expanded size]
+                     * is less than the slaves key cache size * [cacheSizeMultiple],
+                     * then disabled remote streaming.
+                     *
+                     * cacheSizeMultiple because still want to turn off if hitting it e.g. only hitting cache a 1/4 of the time.
+                     *
+                     */
+
+                    return ((size * numParts) * compressionMultiple) < (keyCacheSize * cacheSizeMultiple);
+                };
+
+                if (0 != (helper->getFlags() & TIRusesblob))
+                    tryRemoteStream = false;
+                else if (localMerge)
+                    tryRemoteStream = false;
+                else if (inChildQuery() && !getOptBool(THOROPT_STREAM_REMOTE_SMALL_INDEX)) // option turns of heuristic
```

Comment:
Shouldn't this be:

  if (inChildQuery())
  {
     if (getOptBool())
     {...code below }
     else
          tryRemoteStream = false;

?
Currently the option is forceStreamRemoteSmallIndex, I had assumed it would be allowStreamRemoteSmallIndex from the description.

Perhaps renaming it would make it clear.  Or the option could use the heuristic as the default allowing it to be overridden either way??

Reply:
It is 'forceStreamRemoteSmallIndex' and yes that would be better name for it's current meaning.

I might change it to "minimumCQStreamedIndexSizeMB" , default it to 0, and for that to mean use internal rough heuristic multiples..
If defined, it will be absolute size check, allowing someone to force smaller or larger indexes to be streamed / not-streamed if wanted.